include/kelcoro/thread_pool.hpp

```diff
@@ -184,18 +251,21 @@ struct thread_pool {
   }
 
   explicit thread_pool(size_t thread_count = default_thread_count(),
-                       std::pmr::memory_resource* r = std::pmr::get_default_resource());
+                       std::pmr::memory_resource& r = *std::pmr::get_default_resource())
+      : workers(std::max<size_t>(1, thread_count), r), resource(&r) {
+  }
 
   ~thread_pool() {
-    // if destructor started, then it is undefined behavior to push tasks
-    // because its data race (between destruction and accessing to 'this' for scheduling new task)
-    //
-    // But there is special case - workers, which may invoke task, which schedules next task
-    // in this case, .stop waits for workers consume dead pill, grab all operations from all queues
-    // and cancel them (resume with special errc)
-    // So, no one pushes and all what was pushed by tasks executed on workers is cancelled,
-    // no memory leak, profit!
-    stop(workers, workers_size);
+    task_node pill = task_node::deadpill();
+    for (auto& w : workers) {
+      if (w.thread.joinable()) {
+        w.queue.push(&pill);
+        w.thread.join();
+      }
+    }
+    for (auto& w : workers) {
```

Comment:
also need a comment about why thats not enough to just reset workers (prevent memory leaks / data race / not pushing same pill into all queues because of changing .next, not reseting workers because they all must be stopped and then pushed tasks canceled etc)